File: 0124.BinaryTreeMaximumPathSum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.maxVal = float('-inf')

    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.dfs(root)
        return self.maxVal

    def dfs(self, node):
        if not node:
            return 0
        left = max(0, self.dfs(node.left))
        right = max(0, self.dfs(node.right))
        self.maxVal = max(self.maxVal, left+node.val+right)
        return max(left, right) + node.val

File: 0000-0999/0124.BinaryTreeMaximumPathSum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func maxPathSum(root *TreeNode) int {
    ans := math.MinInt32
    dfs(root, &ans)
    return ans
}

func dfs(node *TreeNode, ans *int) int {
    if node == nil {
        return 0
    }
    maxL := dfs(node.Left, ans)
    maxR := dfs(node.Right, ans)
    *ans = max(*ans, node.Val + max(0, maxL) + max(0, maxR))
    return node.Val + max(0, max(maxL, maxR))
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}